### PR TITLE
Azure network plugin

### PR DESCRIPTION
**Pull request by fgan1**
## Description
Implementing Azure Network Plugin
## Proposed changes
Implement request instance method in the Azure Network plugin
## Additional information
This branch depends on a branch in the Fogbow Common (azure-network-plugin)
## PR Slices
[x] One: Request instance context
[] Two: Get instance context
[] Three: Delete instance context
[] Four: Add
## Reminding
PRs order:
*develop < azure-network-plugin* < PR2 < PR3 < PR4

### CODE REVIEW

File: src/main/java/cloud/fogbow/ras/core/plugins/interoperability/azure/network/sdk/AzureVirtualNetworkOperationSDK.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+                    LOGGER.info(Messages.Info.END_CREATE_VNET_ASYNC_BEHAVIOUR);
+                });
+    }
+
+    @VisibleForTesting
+    Observable<Indexable> buildCreateSecurityGroupObservable(AzureCreateVirtualNetworkRef azureCreateVirtualNetworkRef, Azure azure) {
+        String name = azureCreateVirtualNetworkRef.getName();
+        String resourceGroupName = azureCreateVirtualNetworkRef.getResourceGroupName();
+        String cidr = azureCreateVirtualNetworkRef.getCidr();
+        Region region = Region.fromName(this.regionName);
```
Comment (jadsonluan) [inline src/main/java/cloud/fogbow/ras/core/plugins/interoperability/azure/network/sdk/AzureVirtualNetworkOperationSDK.java]: The **fromName** method may behave differently as expected. This method will search for a region name; if the region is not found, a new region with that name will be created. 

I suggest using **Region.findByLabelOrName** method.

File: src/main/java/cloud/fogbow/ras/core/plugins/interoperability/azure/network/sdk/AzureVirtualNetworkOperationSDK.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    @VisibleForTesting
+    void doNetworkCreationStepTwoSync(Indexable indexableSecurityGroup,
+                                           AzureCreateVirtualNetworkRef azureCreateVirtualNetworkRef,
+                                           Azure azure) {
+
+        NetworkSecurityGroup networkSecurityGroup = (NetworkSecurityGroup) indexableSecurityGroup;
+        String name = azureCreateVirtualNetworkRef.getName();
+        String resourceGroupName = azureCreateVirtualNetworkRef.getResourceGroupName();
+        String cidr = azureCreateVirtualNetworkRef.getCidr();
+        Region region = Region.fromName(this.regionName);
```
Comment (jadsonluan) [inline src/main/java/cloud/fogbow/ras/core/plugins/interoperability/azure/network/sdk/AzureVirtualNetworkOperationSDK.java]: The **fromName** method may behave differently as expected. This method will search for a region name; if the region is not found, a new region with that name will be created. 

I suggest using **Region.findByLabelOrName** method.

File: src/test/java/cloud/fogbow/ras/core/plugins/interoperability/azure/network/sdk/AzureVirtualNetworkOperationSDKTest.java [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+                .thenReturn(observableExpected);
+
+        // exercise
+        Observable<Indexable> observable = this.azureVirtualNetworkOperationSDK.buildCreateSecurityGroupObservable(azureCreateVirtualNetworkRef, this.azure);
+
+        // verify
+        Assert.assertEquals(observableExpected, observable);
+    }
+
+    // test case: When calling the buildVirtualNetworkCreationObservable method and the observables execute
```
Comment (jadsonluan) [inline src/test/java/cloud/fogbow/ras/core/plugins/interoperability/azure/network/sdk/AzureVirtualNetworkOperationSDKTest.java]: The method name present in the test case description doesn't match with the method tested.

Comment (fgan1) [reply src/test/java/cloud/fogbow/ras/core/plugins/interoperability/azure/network/sdk/AzureVirtualNetworkOperationSDKTest.java]: Done.
Note: I have fixed it in the last branch of Azure Network context.

File: src/main/java/cloud/fogbow/ras/core/plugins/interoperability/azure/network/sdk/AzureVirtualNetworkOperationSDK.java (hunk 2 above)
Comment (fgan1) [reply src/main/java/cloud/fogbow/ras/core/plugins/interoperability/azure/network/sdk/AzureVirtualNetworkOperationSDK.java]: Done.
Note: I have fixed it in the last branch of Azure Network context.

File: src/main/java/cloud/fogbow/ras/core/plugins/interoperability/azure/network/sdk/AzureVirtualNetworkOperationSDK.java (hunk 1 above)
Comment (fgan1) [reply src/main/java/cloud/fogbow/ras/core/plugins/interoperability/azure/network/sdk/AzureVirtualNetworkOperationSDK.java]: Done.
Note: I have fixed it in the last branch of Azure Network context.

File: src/main/java/cloud/fogbow/ras/core/plugins/interoperability/azure/network/sdk/AzureNetworkSDK.java [hunk 4]
```diff
@@ -21,8 +28,66 @@
         }
     }
 
+    // TODO(chico) - Implement tests
```
[the remark this answers is not in the document]
Comment (fgan1) [reply src/main/java/cloud/fogbow/ras/core/plugins/interoperability/azure/network/sdk/AzureNetworkSDK.java:31]: Done

File: src/main/java/cloud/fogbow/ras/core/plugins/interoperability/azure/network/sdk/AzureNetworkSDK.java [hunk 5]
```diff
@@ -2,13 +2,20 @@
...
+import rx.Completable;
+import rx.Observable;
 
+import java.util.Map;
 import java.util.Optional;
 
 public class AzureNetworkSDK {
 
+    private static final String DEFAULT_SECURITY_GROUPS_RULES_NAME = "default-security-group-rules-name";
+    private static final String DEFAULT_SUBNET_NAME = "default-subnet-name";
```
[the remark this answers is not in the document]
Comment (fgan1) [reply src/main/java/cloud/fogbow/ras/core/plugins/interoperability/azure/network/sdk/AzureNetworkSDK.java:17]: Done

File: src/main/java/cloud/fogbow/ras/core/plugins/interoperability/azure/network/sdk/AzureNetworkSDK.java [hunk 6]
```diff
@@ -21,8 +28,66 @@
...
+                .define(networkName)
+                .withRegion(region)
+                .withExistingResourceGroup(resourceGroupName)
+                .withTags(tags)
+                .withAddressSpace(cidr)
+                .defineSubnet(DEFAULT_SUBNET_NAME)
+                    .withAddressPrefix(cidr)
+                    .withExistingNetworkSecurityGroup(networkSecurityGroup)
+                    .attach()
+                .create();
```
[the remark this answers is not in the document]
Comment (fgan1) [reply src/main/java/cloud/fogbow/ras/core/plugins/interoperability/azure/network/sdk/AzureNetworkSDK.java:74]: No, I should not. The method says "Sync" that it means Synchronous.

File: src/test/java/cloud/fogbow/ras/core/plugins/interoperability/azure/network/AzureNetworkPluginTest.java [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+import org.powermock.api.mockito.PowerMockito;
+import org.powermock.core.classloader.annotations.PrepareForTest;
+
+import java.io.File;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Properties;
+
+@PrepareForTest({AzureGeneralUtil.class, CreatingInstanceManager.class, CreatingInstanceManager.class})
+public class AzureNetworkPluginTest extends AzureTestUtils {
```
[the remark this answers is not in the document]
Comment (fgan1) [reply src/test/java/cloud/fogbow/ras/core/plugins/interoperability/azure/network/AzureNetworkPluginTest.java:36]: Done

**Pull request merged**